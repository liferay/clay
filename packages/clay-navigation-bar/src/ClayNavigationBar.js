import 'clay-link';
import {Config} from 'metal-state';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';

import templates from './ClayNavigationBar.soy.js';

/**
 * Metal Clay Navbar component.
 */
class ClayNavigationBar extends Component {
	/**
	 * @inheritDoc
	 */
	attached() {
		this.addListener('transitionend', this._handleTransitionEnd, true);
	}

	/**
	 * @inheritDoc
	 */
	// eslint-disable-next-line
	sync_isTransitioning() {
		const elementCollapse = this.element.querySelector('.navbar-collapse');

		if (this._isTransitioning && !this._visible) {
			const heightCollapse = elementCollapse.querySelector('.container')
				.clientHeight;

			elementCollapse.setAttribute(
				'style',
				`height: ${heightCollapse}px`
			);
		} else if (this._isTransitioning && this._visible) {
			elementCollapse.setAttribute('style', 'height: 0');
			elementCollapse.removeAttribute('style');
		}
	}

	/**
	 * Check the click and set transition true.
	 * @private
	 */
	_handleClickToggler() {
		if (!this._isTransitioning) {
			this._isTransitioning = true;
		}
	}

	/**
	 * Handle css transition end.
	 * @param {!Event} event
	 * @private
	 */
	_handleTransitionEnd(event) {
		const element = this.element.querySelector('.navbar-collapse');
		if (
			element == event.target &&
			this._isTransitioning &&
			!this._visible
		) {
			this._visible = true;
			this._isTransitioning = false;
		} else if (element == event.target) {
			this._visible = false;
			this._isTransitioning = false;
		}
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayNavigationBar.STATE = {
	/**
	 * The toggle animation.
	 * @instance
	 * @memberof ClayNavigationBar
	 * @type {?boolean}
	 * @default false
	 * @private
	 */
	_isTransitioning: Config.bool()
		.value(false)
		.internal(),

	/**
	 * Navmenus visible in mobile when click in the button.
	 * @instance
	 * @memberof ClayNavigationBar
	 * @type {?bool}
	 * @default false
	 * @private
	 */
	_visible: Config.bool()
		.value(false)
		.internal(),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayNavigationBar
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayNavigationBar
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * The display style attribute.
	 * @instance
	 * @memberof ClayNavigationBar
	 * @type {?boolean}
	 * @default false
	 */
	inverted: Config.bool().value(false),

	/**
	 * List of items to show in the Navbar.
	 * @instance
	 * @memberof ClayNavigationBar
	 * @type {!Array}
	 * @default undefined
	 */
	items: Config.arrayOf(
		Config.shapeOf({
			active: Config.bool().value(false),
			href: Config.string().required(),
			label: Config.string().required(),
		})
	).required(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayNavigationBar
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),
};

defineWebComponent('clay-navigation-bar', ClayNavigationBar);

Soy.register(ClayNavigationBar, templates);

export {ClayNavigationBar};
export default ClayNavigationBar;
