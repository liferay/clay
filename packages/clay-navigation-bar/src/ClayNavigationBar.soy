{namespace ClayNavigationBar}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param items: list<[
		active: bool,
		href: string,
		label: string
	]>}
	{@param spritemap: string}
	{@param? _handleClickToggler: any}
	{@param? _isTransitioning: bool}
	{@param? _visible: bool}
	{@param? elementClasses: string}
	{@param? id: string}
	{@param? inverted: bool}

	{let $attributes kind="attributes"}
		class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar
			{if $inverted}
				{sp}navigation-bar-secondary
			{else}
				{sp}navigation-bar-light
			{/if}

			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}
		"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	<nav {$attributes}>
		<div class="container-fluid container-fluid-max-xl">
			{if $items}
				{call .toggler}
					{param _handleClickToggler: $_handleClickToggler /}
					{param _visible: $_visible /}
					{param items: $items /}
					{param spritemap: $spritemap /}
				{/call}
			{/if}
			{call .content}
				{param _isTransitioning: $_isTransitioning /}
				{param _visible: $_visible /}
				{param items: $items /}
				{param spritemap: $spritemap /}
			{/call}
		</div>
	</nav>
{/template}

/**
 * This renders a navbar content.
 */
{template .content}
	{@param items: list<[
		active: bool,
		href: string,
		label: string
	]>}
	{@param spritemap: string}
	{@param? _isTransitioning: bool}
	{@param? _visible: bool}

	{let $classes kind="text"}
		navbar-collapse

		{if $_isTransitioning}
			{sp}collapsing
		{else}
			{sp}collapse
		{/if}

		{if $_visible}
			{sp}show
		{/if}
	{/let}

	<div class="{$classes}">
		<div class="container-fluid container-fluid-max-xl">
			{if $items}
				{call .navmenu}
					{param items: $items /}
					{param spritemap: $spritemap /}
				{/call}
			{/if}
		</div>
	</div>
{/template}

/**
 * This renders a navbar item.
 */
{template .navitem}
	{@param href: string}
	{@param label: string}
	{@param? active: bool}

	<li class="nav-item">
		{call ClayLink.render}
			{param ariaLabel kind="text"}
				{if $active}
					Current Page
				{/if}
			{/param}

			{param elementClasses kind="text"}
				nav-link
				{if $active}
					{sp}active
				{/if}
			{/param}

			{param href: $href /}
			{param label kind="html"}
				<span class="navbar-text-truncate">{$label}</span>
			{/param}

		{/call}
	</li>
{/template}

/**
 * This renders list of items.
 */
{template .navmenu}
	{@param items: list<[
		active: bool,
		href: string,
		label: string
	]>}

	<ul class="navbar-nav">
		{foreach $page in $items}
			{call .navitem}
				{param active: $page.active /}
				{param href: $page.href /}
				{param label: $page.label /}
			{/call}
		{/foreach}
	</ul>
{/template}

/**
 * This makes a toggler when viewing on the mobile.
 */
{template .toggler}
	{@param items: list<[
		active: bool,
		href: string,
		label: string
	]>}
	{@param spritemap: string}
	{@param? _handleClickToggler: any}
	{@param? _visible: bool}

	{let $classes kind="text"}
		navbar-toggler navbar-toggler-link
		{if not $_visible}
			{sp}collapsed
		{/if}
	{/let}

	{call ClayLink.render}
		{param ariaExpanded: $_visible /}
		{param ariaLabel: 'Toggle Navigation' /}
		{param elementClasses: $classes /}
		{param events: ['click': $_handleClickToggler] /}
		{param href: '#' /}
		{param icon: 'caret-bottom' /}
		{param iconAlignment: 'right' /}
		{param label kind="text"}
			{foreach $item in $items}
				{if $item.active == true}
					{$item.label}
				{/if}
			{/foreach}
		{/param}

		{param spritemap: $spritemap /}
	{/call}
{/template}
