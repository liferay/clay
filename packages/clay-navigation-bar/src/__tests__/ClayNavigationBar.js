'use strict';

import ClayNavigationBar from '../ClayNavigationBar';

let consoleErrorReference;
let navigationbar;
const spritemap = 'icons.svg';

/**
 * Stubs console.error
 */
function mockConsoleError() {
	console.error = () => {};
}

describe('ClayNavigationBar', function() {
	beforeEach(() => {
		consoleErrorReference = console.error;
	});

	afterEach(() => {
		if (navigationbar) {
			navigationbar.dispose();
		}

		console.error = consoleErrorReference;
	});

	it('should render a navigatiom bar with one element', function() {
		navigationbar = new ClayNavigationBar({
			items: [{label: 'Page 1', href: '#1'}],
			spritemap: spritemap,
		});

		expect(navigationbar).toMatchSnapshot();
	});

	it('should render a navigatiom bar with classes', function() {
		navigationbar = new ClayNavigationBar({
			elementClasses: 'my-custom-class',
			items: [{label: 'Page 1', href: '#1'}],
			spritemap: spritemap,
		});

		expect(navigationbar).toMatchSnapshot();
	});

	it('should render a navigatiom bar with id', function() {
		navigationbar = new ClayNavigationBar({
			id: 'myId',
			items: [{label: 'Page 1', href: '#1'}],
			spritemap: spritemap,
		});

		expect(navigationbar).toMatchSnapshot();
	});

	it('should render an inverted colored navigatiom bar with one element', function() {
		navigationbar = new ClayNavigationBar({
			inverted: true,
			items: [{label: 'Page 1', href: '#1'}],
			spritemap: spritemap,
		});

		expect(navigationbar).toMatchSnapshot();
	});

	it('should render a navigatiom bar with one active element', function() {
		navigationbar = new ClayNavigationBar({
			items: [{active: true, label: 'Page 1', href: '#1'}],
			spritemap: spritemap,
		});

		expect(navigationbar).toMatchSnapshot();
	});

	it('should render a navigatiom bar with more than one element', function() {
		navigationbar = new ClayNavigationBar({
			items: [
				{label: 'Page 1', href: '#1'},
				{label: 'Page 2', href: '#2'},
			],
			spritemap: spritemap,
		});

		expect(navigationbar).toMatchSnapshot();
	});

	it('should fail when no items are passed', function() {
		mockConsoleError();

		expect(() => {
			navigationbar = new ClayNavigationBar({
				spritemap: spritemap,
			});
		}).toThrow();
	});

	it('should fail when no label is passed on each page', function() {
		mockConsoleError();

		expect(() => {
			navigationbar = new ClayNavigationBar({
				items: [{href: '#1'}],
				spritemap: spritemap,
			});
		}).toThrow();
	});

	it('should fail when no href is passed on each page', function() {
		mockConsoleError();

		expect(() => {
			navigationbar = new ClayNavigationBar({
				items: [{label: 'Page 1'}],
				spritemap: spritemap,
			});
		}).toThrow();
	});

	it('should render a navigation bar and emit an event on nav item click', () => {
		navigationbar = new ClayNavigationBar({
			items: [
				{label: 'Page 1', href: '#1'},
				{label: 'Page 2', href: '#2'},
			],
			spritemap: spritemap,
		});

		const dataToExpect = {
			item: {
				label: 'Page 2',
				href: '#2',
			},
		};

		const spy = jest.spyOn(navigationbar, 'emit');

		navigationbar.refs.navItem1.element.click();

		expect(spy).toHaveBeenCalled();
		expect(spy).toHaveBeenCalledWith(
			expect.objectContaining({
				data: dataToExpect,
				name: 'itemClicked',
				originalEvent: expect.any(Object),
			})
		);
	});
});
