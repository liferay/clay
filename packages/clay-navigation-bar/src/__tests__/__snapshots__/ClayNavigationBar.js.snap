// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayNavigationBar should render a navigatiom bar with classes 1`] = `
<nav class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-light my-custom-class">
  <div class="container-fluid container-fluid-max-xl">
    <a aria-expanded="false" aria-label="toggle-navigation" class=" navbar-toggler navbar-toggler-link collapsed" href="#">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom inline-item inline-item-after" focusable="false">
        <title>caret-bottom</title>
        <use xlink:href="icons.svg#caret-bottom"></use>
      </svg>
    </a>
    <div class="navbar-collapse collapse" ref="content">
      <div class="container-fluid container-fluid-max-xl">
        <ul class="navbar-nav">
          <li class="nav-item">
            <a class=" nav-link" href="#1">
              <span class="navbar-text-truncate">Page 1</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`ClayNavigationBar should render a navigatiom bar with id 1`] = `
<nav class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-light" id="myId">
  <div class="container-fluid container-fluid-max-xl">
    <a aria-expanded="false" aria-label="toggle-navigation" class=" navbar-toggler navbar-toggler-link collapsed" href="#">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom inline-item inline-item-after" focusable="false">
        <title>caret-bottom</title>
        <use xlink:href="icons.svg#caret-bottom"></use>
      </svg>
    </a>
    <div class="navbar-collapse collapse" ref="content">
      <div class="container-fluid container-fluid-max-xl">
        <ul class="navbar-nav">
          <li class="nav-item">
            <a class=" nav-link" href="#1">
              <span class="navbar-text-truncate">Page 1</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`ClayNavigationBar should render a navigatiom bar with more than one element 1`] = `
<nav class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-light">
  <div class="container-fluid container-fluid-max-xl">
    <a aria-expanded="false" aria-label="toggle-navigation" class=" navbar-toggler navbar-toggler-link collapsed" href="#">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom inline-item inline-item-after" focusable="false">
        <title>caret-bottom</title>
        <use xlink:href="icons.svg#caret-bottom"></use>
      </svg>
    </a>
    <div class="navbar-collapse collapse" ref="content">
      <div class="container-fluid container-fluid-max-xl">
        <ul class="navbar-nav">
          <li class="nav-item">
            <a class=" nav-link" href="#1">
              <span class="navbar-text-truncate">Page 1</span>
            </a>
          </li>
          <li class="nav-item">
            <a class=" nav-link" href="#2">
              <span class="navbar-text-truncate">Page 2</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`ClayNavigationBar should render a navigatiom bar with n(2) items with id 1`] = `
<nav class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-light">
  <div class="container-fluid container-fluid-max-xl">
    <a aria-expanded="false" aria-label="Toggle Navigation" class=" navbar-toggler navbar-toggler-link collapsed" href="#">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom inline-item inline-item-after" focusable="false">
        <title>caret-bottom</title>
        <use xlink:href="icons.svg#caret-bottom"></use>
      </svg>
    </a>
    <div class="navbar-collapse collapse" ref="content">
      <div class="container-fluid container-fluid-max-xl">
        <ul class="navbar-nav">
          <li class="nav-item">
            <a class=" nav-link" href="#1" id="pageId1">
              <span class="navbar-text-truncate">Page 1</span>
            </a>
          </li>
          <li class="nav-item">
            <a class=" nav-link" href="#2" id="pageId2">
              <span class="navbar-text-truncate">Page 2</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`ClayNavigationBar should render a navigatiom bar with one active element 1`] = `
<nav class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-light">
  <div class="container-fluid container-fluid-max-xl">
    <a aria-expanded="false" aria-label="toggle-navigation" class=" navbar-toggler navbar-toggler-link collapsed" href="#">Page 1
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom inline-item inline-item-after" focusable="false">
        <title>caret-bottom</title>
        <use xlink:href="icons.svg#caret-bottom"></use>
      </svg>
    </a>
    <div class="navbar-collapse collapse" ref="content">
      <div class="container-fluid container-fluid-max-xl">
        <ul class="navbar-nav">
          <li class="nav-item">
            <a aria-label="current-page" class=" nav-link active" href="#1">
              <span class="navbar-text-truncate">Page 1</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`ClayNavigationBar should render a navigatiom bar with one element 1`] = `
<nav class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-light">
  <div class="container-fluid container-fluid-max-xl">
    <a aria-expanded="false" aria-label="toggle-navigation" class=" navbar-toggler navbar-toggler-link collapsed" href="#">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom inline-item inline-item-after" focusable="false">
        <title>caret-bottom</title>
        <use xlink:href="icons.svg#caret-bottom"></use>
      </svg>
    </a>
    <div class="navbar-collapse collapse" ref="content">
      <div class="container-fluid container-fluid-max-xl">
        <ul class="navbar-nav">
          <li class="nav-item">
            <a class=" nav-link" href="#1">
              <span class="navbar-text-truncate">Page 1</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`ClayNavigationBar should render an inverted colored navigatiom bar with one element 1`] = `
<nav class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-secondary">
  <div class="container-fluid container-fluid-max-xl">
    <a aria-expanded="false" aria-label="toggle-navigation" class=" navbar-toggler navbar-toggler-link collapsed" href="#">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom inline-item inline-item-after" focusable="false">
        <title>caret-bottom</title>
        <use xlink:href="icons.svg#caret-bottom"></use>
      </svg>
    </a>
    <div class="navbar-collapse collapse" ref="content">
      <div class="container-fluid container-fluid-max-xl">
        <ul class="navbar-nav">
          <li class="nav-item">
            <a class=" nav-link" href="#1">
              <span class="navbar-text-truncate">Page 1</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;
